src/Microsoft.OData.Core/WriterValidator.cs

```diff
@@ -29,15 +45,50 @@ internal WriterValidator(ODataMessageWriterSettings settings)
             this.settings = settings;
         }
 
-        /// <summary>
-        /// Creates a DuplicatePropertyNameChecker instance.
-        /// </summary>
-        /// <returns>The created instance.</returns>
-        public IDuplicatePropertyNameChecker CreateDuplicatePropertyNameChecker()
+        /// <inheritdoc/>
+        public IDuplicatePropertyNameChecker GetDuplicatePropertyNameChecker()
         {
-            return settings.ThrowOnDuplicatePropertyNames
-                   ? (IDuplicatePropertyNameChecker)new DuplicatePropertyNameChecker()
-                   : (IDuplicatePropertyNameChecker)new NullDuplicatePropertyNameChecker();
+            IDuplicatePropertyNameChecker duplicatePropertyNameChecker;
+
+            if (settings.ThrowOnDuplicatePropertyNames)
+            {
+#if NETSTANDARD2_0_OR_GREATER
+                if (this.duplicatePropertyNameCheckerObjectPool == null)
+                {
+                    DefaultObjectPoolProvider poolProvider = new DefaultObjectPoolProvider { MaximumRetained = 8 };
+                    this.duplicatePropertyNameCheckerObjectPool = poolProvider.Create<DuplicatePropertyNameChecker>();
+                }
+
+                duplicatePropertyNameChecker = this.duplicatePropertyNameCheckerObjectPool.Get();
+                duplicatePropertyNameChecker.Reset();
+#else
+                duplicatePropertyNameChecker = new DuplicatePropertyNameChecker();
+#endif
+            }
+            else
+            {
+                if (this.nullDuplicatePropertyNameChecker == null)
+                {
+                    // NullDuplicatePropertyNameChecker does nothing, so we can use a single instance during the writing process.
+                    this.nullDuplicatePropertyNameChecker = new NullDuplicatePropertyNameChecker();
+                }
+
+                duplicatePropertyNameChecker = this.nullDuplicatePropertyNameChecker;
+            }
+
+            return duplicatePropertyNameChecker;
+        }
+
+        /// <inheritdoc/>
+        public void ReturnDuplicatePropertyNameChecker(IDuplicatePropertyNameChecker duplicatePropertyNameChecker)
+        {
+#if NETSTANDARD2_0_OR_GREATER
+            // We only return the DuplicatePropertyNameChecker to the object pool and ignore the NullDuplicatePropertyNameChecker.
+            if (settings.ThrowOnDuplicatePropertyNames)

```

Comment:
Curious, does this setting guarantee that we're dealing with `DuplicatePropertyNameChecker` and not `NullDuplicatePropertyNameChecker`?

Reply:
#Resolved